<a href="https://colab.research.google.com/github/Kihiko-Peter/Data-Science-Project/blob/main/Moringa_School_Data_Science_Independent_Project_Week_IV_2022_02_PETER_KIHIKO_MAINA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

WEEK IV DATA SCIENCE PROJECT

In [103]:
# Import pandas and numpy
import pandas as pd
import numpy as np
from datetime import datetime

In [104]:
#Loading the dataset
data = pd.read_csv('autolib.csv')

# Creating a dataframe

dataf = pd.DataFrame(data)

In [105]:
# Preview the data
dataf.head(7)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20

[7 rows x 25 columns]

DATA PREPARATION

In [106]:
#Renaming the columns 
#1. Concatenate the year,month and day to be on one column with the datetime object.
dataf['Date']=dataf.apply(lambda x:'%s/%s/%s' % (x['year'],x['month'],x['day']),axis=1)


# Function to confirm dates ate okay
def confirm_date(dataf,Date):
  dataf[Date] = dataf[Date].apply(lambda x :
                                  dataf.datetime.strif(x,'%y/%m/%d'))
dataf.head()



Address  Cars  ...  minute      Date
0                 2 Avenue de Suffren     0  ...      43  2018/4/8
1           145 Rue Raymond Losserand     6  ...      24  2018/4/6
2    2 Avenue John Fitzgerald Kennedy     3  ...      14  2018/4/3
3              51 Rue Eugène Oudiné     3  ...      37  2018/4/4
4  6 avenue de la Porte de Champerret     3  ...      23  2018/4/8

[5 rows x 26 columns]

In [107]:
#Combining hour and minute to be one column time
dataf['Time'] = dataf.apply(lambda a : '%s:%s' % (a['hour'], a['minute']), axis = 1)
dataf.head()

def confirm_date(dataf,Date):
  dataf[Date] = dataf[Date].apply(lambda x :
                                  dataf.datetime.strif(x,'%H : % M'))
dataf.head()



Address  Cars  ...      Date   Time
0                 2 Avenue de Suffren     0  ...  2018/4/8  11:43
1           145 Rue Raymond Losserand     6  ...  2018/4/6   7:24
2    2 Avenue John Fitzgerald Kennedy     3  ...  2018/4/3  20:14
3              51 Rue Eugène Oudiné     3  ...  2018/4/4   4:37
4  6 avenue de la Porte de Champerret     3  ...  2018/4/8  17:23

[5 rows x 27 columns]

DATA CLEANING

Validity - Dealing with the irrelevant data

In [108]:
# Dropping unnecessary columns
#1. Dropping the geo points column since it will have no effect on the study. We have public name
dataf.drop(columns = 'Geo point', inplace = True)
dataf.head()


Address  Cars  ...      Date   Time
0                 2 Avenue de Suffren     0  ...  2018/4/8  11:43
1           145 Rue Raymond Losserand     6  ...  2018/4/6   7:24
2    2 Avenue John Fitzgerald Kennedy     3  ...  2018/4/3  20:14
3              51 Rue Eugène Oudiné     3  ...  2018/4/4   4:37
4  6 avenue de la Porte de Champerret     3  ...  2018/4/8  17:23

[5 rows x 26 columns]

In [109]:
#Dropping the year,month,day , hour and minute column since we have the date time column
dataf.drop(columns = ['year','month', 'day', 'hour', 'minute'], inplace = True)
dataf.head()

Address  Cars  ...      Date   Time
0                 2 Avenue de Suffren     0  ...  2018/4/8  11:43
1           145 Rue Raymond Losserand     6  ...  2018/4/6   7:24
2    2 Avenue John Fitzgerald Kennedy     3  ...  2018/4/3  20:14
3              51 Rue Eugène Oudiné     3  ...  2018/4/4   4:37
4  6 avenue de la Porte de Champerret     3  ...  2018/4/8  17:23

[5 rows x 21 columns]

COMPLETENESS
Checking and handling missing values

In [115]:
#Checking for missing values
def missing(dataf):
  missingu = dataf.isnull.sum()/len(dataf).sort_values(ascending = True)
print(missing)



<function missing at 0x7f8ed83083b0>


In [117]:
#Gives mising values availabl for every column
dataf.isnull().sum()

Address                   0
Cars                      0
Bluecar counter           0
Utilib counter            0
Utilib 1.4 counter        0
Charge Slots              0
Charging Status           0
City                      0
Displayed comment      4889
ID                        0
Kind                      0
Postal code               0
Public name               0
Rental status             0
Scheduled at           4953
Slots                     0
Station type              0
Status                    0
Subscription status       0
Date                      0
Time                      0
dtype: int64

In [118]:
dataf.isnull().any()

Address                False
Cars                   False
Bluecar counter        False
Utilib counter         False
Utilib 1.4 counter     False
Charge Slots           False
Charging Status        False
City                   False
Displayed comment       True
ID                     False
Kind                   False
Postal code            False
Public name            False
Rental status          False
Scheduled at            True
Slots                  False
Station type           False
Status                 False
Subscription status    False
Date                   False
Time                   False
dtype: bool

In [120]:
#Finding the percentage of missing values
percent_missing = dataf.isnull().sum() * 100 / len(dataf)
percent_missing

Address                 0.00
Cars                    0.00
Bluecar counter         0.00
Utilib counter          0.00
Utilib 1.4 counter      0.00
Charge Slots            0.00
Charging Status         0.00
City                    0.00
Displayed comment      97.78
ID                      0.00
Kind                    0.00
Postal code             0.00
Public name             0.00
Rental status           0.00
Scheduled at           99.06
Slots                   0.00
Station type            0.00
Status                  0.00
Subscription status     0.00
Date                    0.00
Time                    0.00
dtype: float64

In [124]:
#Replacing missing values with 0
dataf['Scheduled at'].fillna(0, inplace = True)
dataf['Displayed comment'].fillna(0, inplace = True)
dataf.head()


Address  Cars  ...      Date   Time
0                 2 Avenue de Suffren     0  ...  2018/4/8  11:43
1           145 Rue Raymond Losserand     6  ...  2018/4/6   7:24
2    2 Avenue John Fitzgerald Kennedy     3  ...  2018/4/3  20:14
3              51 Rue Eugène Oudiné     3  ...  2018/4/4   4:37
4  6 avenue de la Porte de Champerret     3  ...  2018/4/8  17:23

[5 rows x 21 columns]

**Conclusion**

99.06% of scheduled at were missing
97.78% of displayed comments were missing

**CONSISTENCY**
Checking for duplicates

In [125]:
#Checking and dropping the duplicates
corrected = dataf.drop_duplicates()
corrected


Address  Cars  ...      Date   Time
0                    2 Avenue de Suffren     0  ...  2018/4/8  11:43
1              145 Rue Raymond Losserand     6  ...  2018/4/6   7:24
2       2 Avenue John Fitzgerald Kennedy     3  ...  2018/4/3  20:14
3                 51 Rue Eugène Oudiné     3  ...  2018/4/4   4:37
4     6 avenue de la Porte de Champerret     3  ...  2018/4/8  17:23
...                                  ...   ...  ...       ...    ...
4995                 8 avenue Ménelotte     2  ...  2018/4/6  11:26
4996                   37 rue de Dantzig     4  ...  2018/4/4  16:56
4997                      142 rue du Bac     1  ...  2018/4/1    7:1
4998         2 avenue du Val de Fontenay     2  ...  2018/4/4  17:27
4999                 5 Rue Henry Brisson     4  ...  2018/4/8   10:2

[5000 rows x 21 columns]

**UNIFORMITY**


In [127]:
#Fixing messy columns
dataf.columns = dataf.columns.str.strip()


DATA **ANALYSIS**

In [157]:
#The most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.
paris = dataf[dataf['City'] == 'Paris']
popular = paris.groupby(['Public name','Bluecar counter'])
pdf =pd.DataFrame(popular)
pdf.head()

#popular['Time'].max()


0                                                  1
0  (Paris/25 Août 1944/8, 0)                         Address  Cars  ...     ...
1  (Paris/25 Août 1944/8, 1)                         Address  Cars  ...     ...
2  (Paris/25 Août 1944/8, 2)                         Address  Cars  ...     ...
3  (Paris/25 Août 1944/8, 3)                        Address  Cars  ...      ...
4  (Paris/25 Août 1944/8, 4)                         Address  Cars  ...     ...

In [149]:
# max number of bluecars
df = dataf['Bluecar counter'].max()
df

7

The highest city with bluecars had 7

In [150]:
# max number of Utilib counter
dataf['Utilib counter'].max()

2

2 was the highest number of utilib counter in the cities

In [153]:
#Mean of cars

dataf['Cars'].mean()

2.3336

Every city had an average of e vehicles